src/Compilers/Core/Portable/Serialization/StreamObjectWriter.cs

```diff
@@ -0,0 +1,1459 @@
+            private readonly List<Variant> _list;
+
+            public VariantListWriter()
+            {
+                _list = new List<Variant>();
+            }
+
+            public List<Variant> List => _list;
+
+            public override void WriteBoolean(bool value)
+            {
+                _list.Add(Variant.FromBoolean(value));
+            }
+
+            public override void WriteByte(byte value)
+            {
+                _list.Add(Variant.FromByte(value));
+            }
+
+            public override void WriteChar(char ch)
+            {
+                _list.Add(Variant.FromChar(ch));
+            }
+
+            public override void WriteDecimal(decimal value)
+            {
+                _list.Add(Variant.FromDecimal(value));
+            }
+
+            public override void WriteDouble(double value)
+            {
+                _list.Add(Variant.FromDouble(value));
+            }
+
+            public override void WriteSingle(float value)
+            {
+                _list.Add(Variant.FromSingle(value));
+            }
+
+            public override void WriteInt32(int value)
+            {
+                _list.Add(Variant.FromInt32(value));
+            }
+
+            public override void WriteInt64(long value)
+            {
+                _list.Add(Variant.FromInt64(value));
+            }
+
+            public override void WriteSByte(sbyte value)
+            {
+                _list.Add(Variant.FromSByte(value));
+            }
+
+            public override void WriteInt16(short value)
+            {
+                _list.Add(Variant.FromInt16(value));
+            }
+
+            public override void WriteUInt32(uint value)
+            {
+                _list.Add(Variant.FromUInt32(value));
+            }
+
+            public override void WriteUInt64(ulong value)
+            {
+                _list.Add(Variant.FromUInt64(value));
+            }
+
+            public override void WriteUInt16(ushort value)
+            {
+                _list.Add(Variant.FromUInt16(value));
+            }
+
+            public override void WriteDateTime(DateTime value)
+            {
+                _list.Add(Variant.FromDateTime(value));
+            }
+
+            public override void WriteString(string value)
+            {
+                if (value == null)
+                {
+                    _list.Add(Variant.Null);
+                }
+                else
+                {
+                    _list.Add(Variant.FromString(value));
+                }
+            }
+
+            public override void WriteValue(object value)
+            {
+                _list.Add(Variant.FromBoxedObject(value));
+            }
+        }
+
+        private class ReferenceMap : IDisposable
+        {
+            internal static readonly ObjectPool<Dictionary<object, int>> DictionaryPool =
+                new ObjectPool<Dictionary<object, int>>(() => new Dictionary<object, int>(128), 2);
```

Comment:
Is that the behavior we want though?  Since this is a 'reference map', i assumed we wanted *reference equality*.